src/approaches/pg3_approach.py

```diff
@@ -0,0 +1,429 @@
+        if self._ldl_solves_abstract_task(ldl, atoms, goal):
+            return 0.0
+        return 1.0
+
+    @staticmethod
+    def _ldl_solves_abstract_task(ldl: LiftedDecisionList,
+                                  atoms: Set[GroundAtom],
+                                  goal: Set[GroundAtom]) -> bool:
+        for _ in range(CFG.horizon):
+            if goal.issubset(atoms):
+                return True
+            ground_nsrt = utils.query_ldl(ldl, atoms, goal)
+            if ground_nsrt is None:
+                return False
+            atoms = utils.apply_operator(ground_nsrt, atoms)
+        return goal.issubset(atoms)
+
+
+class _PlanComparisonPG3Heuristic(_PG3Heuristic):
+    """Score a policy based on agreement with certain plans."""
+
+    def _get_score_for_task(self, ldl: LiftedDecisionList,
+                            task_idx: int) -> float:
+        atom_plan = self._get_atom_plan_for_task(ldl, task_idx)
+        # Note: we need the goal because it's an input to the LDL policy.
+        _, _, goal = self._abstract_train_tasks[task_idx]
+        assert goal.issubset(atom_plan[-1])
+        return self._count_missed_steps(ldl, atom_plan, goal)
+
+    @abc.abstractmethod
+    def _get_atom_plan_for_task(self, ldl: LiftedDecisionList,
+                                task_idx: int) -> Sequence[Set[GroundAtom]]:
+        """Given a task, get the plan with which we will compare the policy."""
+        raise NotImplementedError("Override me!")
+
+    @staticmethod
+    def _count_missed_steps(ldl: LiftedDecisionList,
+                            atom_seq: Sequence[Set[GroundAtom]],
+                            goal: Set[GroundAtom]) -> float:
+        missed_steps = 0.0
+        for t in range(len(atom_seq) - 1):
+            ground_nsrt = utils.query_ldl(ldl, atom_seq[t], goal)
+            if ground_nsrt is None:
+                missed_steps += CFG.pg3_plan_compare_inapplicable_cost
+            else:
+                predicted_atoms = utils.apply_operator(ground_nsrt,
+                                                       atom_seq[t])
+                if predicted_atoms != atom_seq[t + 1]:
+                    missed_steps += 1
+        return missed_steps
+
+
+class _DemoPlanComparisonPG3Heuristic(_PlanComparisonPG3Heuristic):
+    """Score a policy based on agreement with demo plans.
+
+    This heuristic assumes that there is exactly one demonstration per
+    task.
+    """
+
+    def _get_atom_plan_for_task(self, ldl: LiftedDecisionList,
+                                task_idx: int) -> Sequence[Set[GroundAtom]]:
+        del ldl  # unused
+        return self._get_demo_atom_plan_for_task(task_idx)
+
+    @functools.lru_cache(maxsize=None)
+    def _get_demo_atom_plan_for_task(
+            self, task_idx: int) -> Sequence[Set[GroundAtom]]:
+        demo_atom_plan = None
+        for ground_atom_traj in self._ground_atom_trajs:
+            traj, atom_seq = ground_atom_traj
```

Comment:
generally in our codebase we use `atoms_seq` instead of `atom_seq` -- maybe do a find-and-replace?